File: sentimental-credit/credit.py
from cs50 import get_int


def math(n_str):

    i = len(n_str) - 2
    m_1 = 0
    m_2 = 0

    while i >= 0:

        m1 = str(2 * int(n_str[i]))
        m2 = str(int(n_str[i+1]))
        m_1 += sum(list(int(x) for x in m1))
        m_2 += sum(list(int(x) for x in m2))
        if i == 1:
            m_2 += int(n_str[0])
        i -= 2

    if ((m_1 + m_2) % 10) == 0:

        return True


def card():

    while True:

        number = get_int("Number: ")
        n_str = str(number)

        if math(n_str):

            if n_str[0:2] in ["34", "37"] and len(n_str) == 15:
                print("AMEX\n")
                return False

            elif n_str[0:2] in ["51", "52", "53", "54", "55"] and len(n_str) == 16:
                print("MASTERCARD\n")
                return False

            elif n_str[0] in ["4"] and len(n_str) in [13, 16]:
                print("VISA\n")
                return False

            else:
                print("INVALID\n")
                return False

        else:
            print("INVALID\n")
            return False


card()

File: C/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int luhn_algo(long n);

int count(long n);

int amex_mast(long n);

int visa(long n);

int main(void)
{
    long n = get_long("Credit card number: ");

    if ((luhn_algo(n) % 10) == 0)
    {
        if ((visa(n) == 4) && (count(n) == 13 || count(n) == 16))
        {
            printf("VISA\n");
        }
        else if ((amex_mast(n) == 51 || amex_mast(n) == 52 || amex_mast(n) == 53 ||
                  amex_mast(n) == 54 || amex_mast(n) == 55) &&
                 (count(n) == 16))
        {
            printf("MASTERCARD\n");
        }
        else if ((amex_mast(n) == 34 || amex_mast(n) == 37) && (count(n) == 15))
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int luhn_algo(long n)
{
    long m = n / 10;
    int step = pow(10, 2);
    int sum = 0;

    while (m > 0)
    {
        int digit = (m % 10) * 2;

        while (digit > 0)
        {
            sum += digit % 10;
            digit /= 10;
        }

        m /= step;
    }

    while (n > 0)
    {
        sum += n % 10;
        n /= step;
    }
    return sum;
}

int count(long n)
{
    int i = 0;
    while (n > 0)
    {
        i++;
        n /= 10;
    }
    return i;
}

int amex_mast(long n)
{
    while (n >= 100)
    {
        n /= 10;
    }
    return n;
}

int visa(long n)
{
    while (n >= 10)
    {
        n /= 10;
    }
    return n;
}
